runtime/commonMain/src/kotlinx/serialization/Coders.kt

```diff
@@ -125,31 +125,130 @@ interface Decoder {
 
     /**
-     * Optional method to specify collection size to pre-allocate memory.
-     * If decoder specifies stream reading ([READ_ALL] is returned from [decodeElementIndex], then
-     * correct implementation of this method is mandatory.
+     *  Decodes the index of the next element to be decoded.
+     *  Index represents a position of the current element in the serial descriptor element that can be found
+     *  with [SerialDescriptor.getElementIndex].
+     *
+     *  If this method returns non-negative index, the caller should call one of the `decode*Element` methods
+     *  with a resulting index.
+     *  Apart from positive values, this method can return [READ_DONE] to indicate that no more elements
+     *  are left or [UNKNOWN_NAME] to indicate that symbol with an unknown name was encountered.
+     *
+     * Example of usage:
+     * ```
+     * class MyPair(i: Int, d: Double)
+     *
+     * object MyPairSerializer : KSerializer<MyPair> {
+     *     // ... other methods omitted
+     *
+     *    fun deserialize(decoder: Decoder): MyPair {
+     *        val composite = decoder.beginStructure(descriptor)
+     *        var i: Int? = null
+     *        var d: Double? = null
+     *        while (true) {
+     *            val index = composite.decodeElementIndex(descriptor)
+     *            if (index == READ_DONE) break // Input is over
+     *            when (index) {
+     *                0 -> {
+     *                    i = composite.decodeIntElement(descriptor, 0)
+     *                }
+     *                1 -> {
+     *                    d = composite.decodeDoubleElement(descriptor, 1)
+     *                }
+     *                else -> error("Unexpected index: $index)
+     *            }
+     *
+     *        }
+     *        composite.endStructure(descriptor)
+     *        if (i == null || d == null) throwMissingFieldException()
+     *        return MyPair(i, d)
+     *    }
+     * }
+     * ```
+     * This example is a rough equivalent of what serialization plugin generates for serializable pair class.
      *
-     * @return Collection size or -1 if not available.
+     * The need in such loop comes from unstructured nature of most serialization formats.
+     * For example, JSON for the following input `{"d": 2.0, "i": 1}`, will first read `d` key with index `1`
+     * and only after `i` with index `0`.
+     *
+     * A potential implementation of this method for JSON format can be the following:
+     * ```
+     * fun decodeElementIndex(descriptor: SerialDescriptor): Int {
+     *     // Ignore arrays
+     *     val nextKey: String? = myStringJsonParser.nextKey()
+     *     if (nextObjectKey == null) return READ_DONE
+     *     return descriptor.getElementIndex(nextKey) // getElementIndex can return UNKNOWN_FIELD
+     * }
+     * ```
      */
-    fun decodeCollectionSize(desc: SerialDescriptor): Int = -1
+    public fun decodeElementIndex(descriptor: SerialDescriptor): Int
```

Comment:
I would also like to extract Composite* into its own file and rename `desc` parameter to `descriptor` as part of this PR, if you don't mind